cstore_reader.c

```diff
@@ -1291,76 +1359,70 @@ ColumnDefaultValue(TupleConstr *tupleConstraints, Form_pg_attribute attributeFor
 }
 
 
-/* Returns the size of the given file handle. */
-static int64
-FileSize(FILE *file)
+/* ReadFromFile reads the given segment from the relation data file. */
+static StringInfo
+ReadFromFile(Relation relation, uint64 offset, uint32 size)
 {
-	int64 fileSize = 0;
-	int fseekResult = 0;
-
-	errno = 0;
-	fseekResult = fseeko(file, 0, SEEK_END);
-	if (fseekResult != 0)
-	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not seek in file: %m")));
-	}
+	int blockCapacity = CSTORE_PAGE_DATA_SIZE;
+	uint32 blockNumber = offset / blockCapacity;
+	uint32 blockOffset = offset % blockCapacity;
+	uint32 remainingSize = size;
+	uint32 resultOffset = 0;
+	uint32 blockCount = 0;
+	StringInfo resultBuffer = NULL;
+	BufferAccessStrategy strategy = GetAccessStrategy(BAS_BULKREAD);
 
-	fileSize = ftello(file);
-	if (fileSize == -1)
+	if (size == 0)
 	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not get position in file: %m")));
+		return NULL;
 	}
 
-	return fileSize;
-}
-
-
-/* Reads the given segment from the given file. */
-static StringInfo
-ReadFromFile(FILE *file, uint64 offset, uint32 size)
-{
-	int fseekResult = 0;
-	int freadResult = 0;
-	int fileError = 0;
-
-	StringInfo resultBuffer = makeStringInfo();
+	resultBuffer = makeStringInfo();
 	enlargeStringInfo(resultBuffer, size);
-	resultBuffer->len = size;
 
-	if (size == 0)
+	blockCount = RelationGetNumberOfBlocksInFork(relation, DATA_FORKNUM);
+	if (blockNumber > blockCount)
 	{
-		return resultBuffer;
+		ereport(ERROR, (errcode(ERRCODE_FDW_ERROR),
+						errmsg("could not read from file")));
 	}
```

Comment:
Hm, isn't this just a repeat of a check that postgres also does?

Reply:
This felt easier to debug. Postgres' check is buried deep down and difficult to follow for an untrained eye.

Reply:
-1

Reply:
But that means the code is unreachable, and you'll get a more descriptive error first. It's better to add a comment saying that it's "guaranteed" to succeed or never reach that part of the code.

Reply:
replace this with asserd